### PR TITLE
onionshare: 2.3.3 -> 2.4

### PR DESCRIPTION
**Pull request by bbjubjub2494**
###### Motivation for this change
https://github.com/onionshare/onionshare/releases/tag/v2.4

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lourkeur** [comment]
Something's wrong (`onionshare «some file to share»`)
```
Connecting to the Tor network: 100% - Done
Your version of Tor is too old, stealth onion services are not supported
```

**Comment by lourkeur** [comment]
> Your version of Tor is too old, stealth onion services are not supported

https://github.com/onionshare/onionshare/blob/cf9dfad1c62e826dfbe112566571a6600d51c230/cli/onionshare_cli/onion.py#L588

**Comment by lourkeur** [comment]
> > Your version of Tor is too old, stealth onion services are not supported
> 
> https://github.com/onionshare/onionshare/blob/cf9dfad1c62e826dfbe112566571a6600d51c230/cli/onionshare_cli/onion.py#L588

```
TypeError: create_ephemeral_hidden_service() got an unexpected keyword argument 'client_auth_v3'
```

That's because upstream uses a custom version of the `stem` library.
https://github.com/onionshare/onionshare/blob/5de48bba6e3c17b0b45516fe536a3c500fab4fc4/cli/pyproject.toml#L32

**Comment by lourkeur** [comment]
function tested :ballot_box_with_check:

**Comment by risicle** [comment]
https://github.com/onionshare/stem/archive/1.8.1-onionshare.tar.gz 404ing for me.

**Comment by lourkeur** [comment]
> https://github.com/onionshare/stem/archive/1.8.1-onionshare.tar.gz 404ing for me.

my bad! fixed

File: pkgs/applications/networking/onionshare/stem.nix [hunk 1]
```diff
@@ -0,0 +1,34 @@
+{ lib, buildPythonPackage, fetchFromGitHub, python, mock }:
+
+buildPythonPackage rec {
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/onionshare/stem.nix:3]: Shouldn't this be under pythonPackages?

Comment (lourkeur) [reply pkgs/applications/networking/onionshare/stem.nix:3]: It should be and it is but Onionshare depends on its customized fork of that module. The normal stem from Tor Project doesn't work. (for now)

Comment (risicle) [reply pkgs/applications/networking/onionshare/stem.nix:3]: Curious why you didn't override the existing package for this though.,

**Comment by risicle** [comment]
`nixpkgs-review` happy, nixos x86_64. Seems to work.

On macos 10.15 just a single test is failing which is annoying, but nothing new:

```
>           assert webhook_url == "http://127.0.0.1:1337/example"
E           AssertionError: assert None == 'http://127.0.0.1:1337/example'

tests/test_cli_web.py:166: AssertionError
----------------------------- Captured stdout call -----------------------------
Sep 28, 10:16PM: Upload of total size 269.0 B is starting
Could not create OnionShare data folder: /var/empty/OnionShare/2021-09-28/221629
=========================== short test summary info ============================
FAILED tests/test_cli_web.py::TestWeb::test_receive_mode_webhook - AssertionE..
```

**Comment by lourkeur** [comment]
> `nixpkgs-review` happy, nixos x86_64. Seems to work.
> 
> On macos 10.15 just a single test is failing which is annoying, but nothing new:
> 
> ```
> >           assert webhook_url == "http://127.0.0.1:1337/example"
> E           AssertionError: assert None == 'http://127.0.0.1:1337/example'
> 
> tests/test_cli_web.py:166: AssertionError
> ----------------------------- Captured stdout call -----------------------------
> Sep 28, 10:16PM: Upload of total size 269.0 B is starting
> Could not create OnionShare data folder: /var/empty/OnionShare/2021-09-28/221629
> =========================== short test summary info ============================
> FAILED tests/test_cli_web.py::TestWeb::test_receive_mode_webhook - AssertionE..
> ```

should I disable that test?

File: pkgs/applications/networking/onionshare/stem.nix (hunk 1 above)
Comment (lourkeur) [reply pkgs/applications/networking/onionshare/stem.nix]: I thought it wouldn't work, but actually it does so there you go.

File: pkgs/applications/networking/onionshare/default.nix [hunk 2]
```diff
@@ -12,6 +12,7 @@
   psutil,
   pyqt5,
   pycrypto,
+  pynacl,
   pyside2,
   pytestCheckHook,
   qrcode,
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/onionshare/default.nix:18]:
```suggestion
, psutil
, pyqt5
, pycrypto
, pynacl
, pyside2
, pytestCheckHook
, qrcode
```

File: pkgs/applications/networking/onionshare/default.nix [hunk 3]
```diff
@@ -1,26 +1,26 @@
...
+, pynacl
+, pyside2
+, pytestCheckHook
+, qrcode
+, qt5
+, requests
+, unidecode
+, tor
+, obfs4
+,
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/onionshare/default.nix:23]:
```suggestion
```

**Pull request merged**